source/common/quic/quic_transport_socket_factory.h

```diff
@@ -114,13 +116,22 @@ class QuicClientTransportSocketFactory : public QuicTransportSocketFactoryBase {
     return fallback_factory_->config();
   }
 
+  // Returns a crypto config generated from the up-to-date client context config. Once the passed in
+  // context config gets updated, a new crypto config object will be returned by this method.
+  std::shared_ptr<quic::QuicCryptoClientConfig> getCryptoConfig();
+
 protected:
   // fallback_factory_ will update the context.
   void onSecretUpdated() override {}
```

Comment:
can't we update here by implementing onSecretUpdated, instead of latching and comparing pointers?

Reply:
I think we used the lazy update because it was a bit tricky to add onSecretUpdated() to the callback list. Because fallback_factory_ provides the sslCtx object. We need to register this callback after the fallback_factory_ registered itself, which is during its construction, to ensure when onSecretUpdated() is triggered there is an updated sslCtx from the fallback_factory_ to use. But because fallback_factory_ owns the config object. After its construction, there is no non-const handle to the config object for this class to register onSecretUpdated(). 

I could have added a member variable to point to config object before passing it to fallback_factory_, but such handle short-lives the base class. So it's not a good model.